### PR TITLE
Allow console connection to paused VM

### PR DESCRIPTION
**Pull request by andreyod**
**What this PR does / why we need it**:

Until now console connection to paused VM was not permitted. This PR will allow serial and VNC console into paused VM. Console will be active after VM is unpaused. 
This can be useful for debugging. In the future we can start VM with it's guest paused and debug guest OS boot issues.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/rest/subresource.go
Comment (iholder-redhat) [inline pkg/virt-api/rest/subresource.go:176]: Do we want also VNC connections to establish for paused VMs? What's the use case for that?

File: pkg/virtctl/templates/templates.go
Comment (iholder-redhat) [inline pkg/virtctl/templates/templates.go:68]: I would avoid the comment and instead put all the information in the function name, e.g. maybe change `WarnPaused` to something like `PrintWarnningForPausedVMI`?

File: pkg/virtctl/templates/templates.go
Comment (iholder-redhat) [inline pkg/virtctl/templates/templates.go:69]: I would also consider changing `name` to `vmiName` to be more clear

File: tests/pausing_test.go
Comment (iholder-redhat) [inline tests/pausing_test.go:338]: I can't comment on the right line, I meant line 343.

You can also test that the warning message is indeed showing up.
Something like:
```
command := tests.NewVirtctlCommand("pause", "vm", "--namespace", tests.NamespaceTestDefault, vm.Name)
out := command.InOrStdin()
scanner := bufio.NewScanner(out)
scanner.Buffer(make([]byte, 1024), 512*1024)
for scanner.Scan() {
	// Ensure your message was printed
}
```

File: tests/pausing_test.go
Comment (iholder-redhat) [inline tests/pausing_test.go:352]: Same in this test

File: pkg/virt-api/rest/subresource_test.go
Comment (iholder-redhat) [inline pkg/virt-api/rest/subresource_test.go:318]: Maybe it's better to keep the test and to expect that the the connection will succeed? Exactly like you did for functional tests.

File: pkg/virt-api/rest/subresource_test.go
Comment (iholder-redhat) [inline pkg/virt-api/rest/subresource_test.go:345]: same as above

**Review by iholder-redhat — COMMENTED** [review]
Looks great! Thanks!!

File: pkg/virt-api/rest/subresource.go
Comment (andreyod) [reply pkg/virt-api/rest/subresource.go:176]: User can choose between serial and VNC console. So we want both.

File: pkg/virtctl/templates/templates.go
Comment (andreyod) [reply pkg/virtctl/templates/templates.go]: Changed the function name. Kept the comment as doc comment for exported function.

File: pkg/virtctl/templates/templates.go
Comment (andreyod) [reply pkg/virtctl/templates/templates.go]: Changed.

File: tests/pausing_test.go
Comment (andreyod) [reply tests/pausing_test.go:338]: The warning message is printed on virtctl console command, not on pause. I'm not sure it is feasible to automate virtctl console command.

File: pkg/virt-api/rest/subresource_test.go
Comment (andreyod) [reply pkg/virt-api/rest/subresource_test.go:345]: Those unit tests was testing specific code in subresource.go handlers. I removed that code so I think it's ok to remove the unit tests. Handler functions are still covered by other tests in subresource_test.go

File: tests/pausing_test.go
Comment (iholder-redhat) [reply tests/pausing_test.go:338]: Oh, got you.
I'm not sure it's feasible too. Thanks!

File: pkg/virt-api/rest/subresource_test.go
Comment (iholder-redhat) [reply pkg/virt-api/rest/subresource_test.go:345]: I see your point.
I personally think that it's better to keep the tests to ensure that it's possible to connect to paused VMs. After all you didn't delete code because it is not relevant but to allow a new capability which IMO needs to be checked in tests.

On the other hand it is being checked via functional tests so I wouldn't say it's a must. Your call :)

**Comment by iholder-redhat** [comment]
Thanks!
/lgtm

File: tests/pausing_test.go
Comment (enp0s3) [inline tests/pausing_test.go:363]: @andreyod Can you please add an expecter to test the console output of `PrintWarningForPausedVMI`?

File: tests/pausing_test.go
Comment (enp0s3) [inline tests/pausing_test.go:338]: Please don't forget to update Polarion with the new headline as soon as the PR will be merged.

File: tests/pausing_test.go
Comment (enp0s3) [inline tests/pausing_test.go:352]: same here.

**Review by enp0s3 — CHANGES_REQUESTED** [review]

File: tests/pausing_test.go
Comment (enp0s3) [reply tests/pausing_test.go:363]: @andreyod Actually I am not sure if it's possible to catch the warning output on the console stream IO. I think maybe it's worth to create a K8S event of when VMI is paused.

Comment (andreyod) [reply tests/pausing_test.go:363]: Yes. We can't catch the warning here as it is in virtctl and not on console stream IO

Comment (andreyod) [reply tests/pausing_test.go:363]: @enp0s3  So you propose to create K8S event when client is trying to open console to paused VM?

File: tests/pausing_test.go
Comment (andreyod) [reply tests/pausing_test.go:338]: Will do

File: tests/pausing_test.go
Comment (andreyod) [reply tests/pausing_test.go:352]: Will do

File: tests/pausing_test.go
Comment (enp0s3) [reply tests/pausing_test.go:363]: @andreyod Let's first check whether the UI is capable of checking the VMI status condition. If so, the event won't be needed since the front-end can react to the condition. Otherwise let's propose to add an event

Comment (andreyod) [reply tests/pausing_test.go:363]: Yes. UI is capable of checking the VMI status condition so don't need an event.

**Review by enp0s3 — APPROVED** [review]
Thank you!
/approve

**Pull request merged**
